azext_iot/monitor/property.py

```diff
@@ -7,91 +7,202 @@
 import datetime
 import isodate
 import time
-
+from azext_iot.monitor.parsers import strings
+from azext_iot.monitor.models.enum import Severity
 from azext_iot.constants import (
     CENTRAL_ENDPOINT,
     DEVICETWIN_POLLING_INTERVAL_SEC,
     DEVICETWIN_MONITOR_TIME_SEC,
+    PNP_INTERFACE_PREFIX,
 )
-from azext_iot.central.providers.devicetwin_provider import CentralDeviceTwinProvider
-from azext_iot.central.models.devicetwin import DeviceTwin, Property
-
-
-def start_property_monitor(
-    cmd,
-    device_id,
-    app_id,
-    central_dns_suffix=CENTRAL_ENDPOINT,
-    polling_interval_seconds=DEVICETWIN_POLLING_INTERVAL_SEC,
-):
-    prev_twin = None
-
-    device_twin_provider = CentralDeviceTwinProvider(
-        cmd=cmd, app_id=app_id, device_id=device_id
-    )
 
-    while True:
-
-        raw_twin = device_twin_provider.get_device_twin(
-            central_dns_suffix=central_dns_suffix
+from azext_iot.central.models.devicetwin import DeviceTwin, Property
+from azext_iot.central.providers import (
+    CentralDeviceProvider,
+    CentralDeviceTemplateProvider,
+    CentralDeviceTwinProvider,
+)
+from azext_iot.monitor.parsers.issue import IssueHandler
+
+
+class PropertyMonitor:
+    def __init__(
+        self, cmd, app_id, device_id, central_dns_suffix=CENTRAL_ENDPOINT,
+    ):
+        self._cmd = cmd
+        self._app_id = app_id
+        self._device_id = device_id
+        self._central_dns_suffix = central_dns_suffix
+        self._template = self._get_device_template()
+        self._issues_handler = IssueHandler()
+
+    def _compare_properties(self, prev_prop: Property, prop: Property):
+        if prev_prop.version == prop.version:
+            return
+
+        changes = {
+            key: self._changed_props(prop.props[key], prop.metadata[key], key,)
+            for key, val in prop.metadata.items()
+            if self._is_relevant(key, val)
+        }
+
+        return changes
+
+    def _is_relevant(self, key, val):
+        if key in {"$lastUpdated", "$lastUpdatedVersion"}:
+            return False
+
+        updated_within = datetime.datetime.now() - datetime.timedelta(
+            seconds=DEVICETWIN_MONITOR_TIME_SEC
```

Comment:
nit - might be worth making this configurable (i.e. use a function param that defaults to this value instead)

Reply:
wont fix